core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/variable/listener/support/VariableListenerSupport.java

```diff
@@ -76,27 +61,11 @@ public void linkVariableListeners() {
                         // Non-sourced variable listeners (ie. ones provided by the user) can never be a supply.
                         supplyMap.put(shadowVariableDescriptor.getProvidedDemand(), (Supply) variableListener);
                     }
-                    int globalOrder = shadowVariableDescriptor.getGlobalShadowOrder();
-                    if (nextGlobalOrder <= globalOrder) {
-                        nextGlobalOrder = globalOrder + 1;
-                    }
-                    VariableListenerNotifiable<Solution_> notifiable =
-                            new VariableListenerNotifiable<>(variableListener, globalOrder);
-                    for (VariableDescriptor<Solution_> source : shadowVariableDescriptor.getSourceVariableDescriptorList()) {
-                        List<VariableListenerNotifiable<Solution_>> variableNotifiableList =
-                                sourceVariableToNotifiableMap.get(source);
-                        variableNotifiableList.add(notifiable);
-                        List<VariableListenerNotifiable<Solution_>> entityNotifiableList = sourceEntityToNotifiableMap
-                                .get(source.getEntityDescriptor());
-                        if (!entityNotifiableList.contains(notifiable)) {
-                            entityNotifiableList.add(notifiable);
-                        }
-                    }
-                    notifiableList.add(notifiable);
+                    notifiableRegistry.registerShadowVariableListener(shadowVariableDescriptor, variableListener);
                 }
             }
         }
-        Collections.sort(notifiableList);
+        notifiableRegistry.sort();
```

Comment:
Suggestion: make all registrations in a batch, then you don't need to `sort()` externally. One less thing to remember, one less thing to introduce bugs.

Reply:
Batch registration cannot be done easily but yes, this could be improved. Good point, I'm working on it.